Examples/Io/Performance/ActsExamples/Io/Performance/VertexPerformanceWriter.cpp

```diff
@@ -243,219 +242,175 @@ ActsExamples::ProcessCode ActsExamples::VertexPerformanceWriter::writeT(
-        }
+    auto iparticle = particles.find(particleMatch.particle->value());
+    if (iparticle == particles.end()) {
+      ACTS_DEBUG(
+          "Truth particle found but not monitored with this track, index = "
+          << track.index() << " tip index = " << track.tipIndex()
+          << " and this barcode = " << particleMatch.particle->value());
+      return {};
+    }
 
-        auto it = std::find_if(allTruthParticles.begin(),
-                               allTruthParticles.end(), [&](const auto& tp) {
-                                 return tp.particleId() == majorityParticleId;
-                               });
+    return *iparticle;
+  };
 
-        if (it == allTruthParticles.end()) {
-          continue;
-        }
+  if (m_cfg.useTracks) {
+    tracks = &m_inputTracks(ctx);
 
-        trackParameters.push_back(track.createParametersAtReference());
-        const auto& majorityParticle = *it;
-        associatedTruthParticles.push_back(majorityParticle);
+    for (const auto& track : *tracks) {
+      if (!track.hasReferenceSurface()) {
+        ACTS_DEBUG("No reference surface on this track, index = "
+                   << track.index() << " tip index = " << track.tipIndex());
+        continue;
       }
-    }
 
-    if (trackParameters.size() != associatedTruthParticles.size()) {
-      ACTS_ERROR(
-          "Number of track parameters and associated truth particles do not "
-          "match. ("
-          << trackParameters.size() << " != " << associatedTruthParticles.size()
-          << ").");
+      if (auto particle = findParticle(track); particle.has_value()) {
+        recoParticles.insert(*particle);
+      }
     }
   } else {
-    // if not using tracks, then all truth particles
-    // are associated with the vertex
-    associatedTruthParticles =
-        std::vector<SimParticle>(m_inputAllTruthParticles(ctx).begin(),
-                                 m_inputAllTruthParticles(ctx).end());
+    // if not using tracks, then all truth particles are associated with the
+    // vertex
+    recoParticles = particles;
   }
 
   // Get number of track-associated true primary vertices
-  m_nVtxReconstructable =
-      getNumberOfReconstructableVertices(SimParticleContainer(
-          associatedTruthParticles.begin(), associatedTruthParticles.end()));
-
-  ACTS_INFO(
-      "Total number of reconstructed tracks : " << trackParameters.size());
-  ACTS_INFO("Total number of reco track-associated truth particles in event : "
-            << associatedTruthParticles.size());
-  ACTS_INFO("Total number of reco track-associated truth primary vertices : "
-            << m_nVtxReconstructable);
+  m_nVtxReconstructable = getNumberOfReconstructableVertices(recoParticles);
 
-  // Loop over all reco vertices and find associated truth particles
-  std::vector<SimParticleContainer> truthParticlesAtVtxContainer;
+  ACTS_INFO("Number of reconstructed tracks : "
+            << ((tracks != nullptr) ? tracks->size() : 0));
```

Comment:
This whole printout doesn't make any sense if no tracks are given, right?

Reply:
Right I guess it is just a signal to confirm that we do not have any tracks